File: server/src/lua/example_dataspace.lua
package.cpath = '/rom/?.so;' .. package.cpath

local util = require 'util'

local L4Re = {}

local function load_l4re_module(name)
  local module = require(name)

  if (module.L4Re ~= nil) then
    for k, v in pairs(module.L4Re) do
      L4Re[k] = v
    end
  end
end

load_l4re_module('capability')
load_l4re_module('cap_alloc')
load_l4re_module('dataspace')
load_l4re_module('env')
load_l4re_module('mem_alloc')
load_l4re_module('rm')

local str =
  "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod " ..
  "tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At " ..
  "vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd " ..
  "gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum " ..
  "dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor " ..
  "invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero " ..
  "eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no " ..
  "sea takimata sanctus est Lorem ipsum dolor sit amet."


local function alloc(ma, size)
  local cap_alloc = L4Re.Util.get_cap_alloc()

  local ds = cap_alloc:alloc_dataspace()

  assert(ma:access():alloc(size, ds) == 0)

  assert(ds:access():allocate(0, size) == 0)

  return ds
end

local function reserve(rm, size)
  local start = 0
  err, start = rm:access():reserve_area(start,
                                        size,
                                        L4Re.RM_F_SEARCH_ADDR)
  assert(err == 0)

  return start
end

local function map(ds, start, flags)
  assert(ds:access():map_region(0, flags, start, start + ds:access():size()) == 0)
end

local function attach(rm, ds, flags)
  local start = 0

  err, start = rm:access():attach(start,
                                  ds:access():size(),
                                  L4Re.RM_F_SEARCH_ADDR | flags,
                                  ds)
  assert(err == 0)

  return start
end

-- environment
local env = L4Re.Env.env()

local ma = env:mem_alloc_1()
local rm = env:rm_1()

-- create dataspace
local ds = alloc(ma, #str)

local ds_size = ds:access():size()
assert(ds_size == #str)

local ds_flags = ds:access():flags()
local ds_rights = ds_flags:fpage_rights()
assert(ds_rights == L4Re.DATASPACE_F_RWX)

-- map dataspace
local ds_start = reserve(rm, ds_size)

map(ds, ds_start, L4Re.DATASPACE_F_RW)

util.strcpy(ds_start, str)

assert(util.strcmp(ds_start, str) == 0)

-- clear dataspace
ds:access():clear(0, ds_size)

local zeros = string.rep("\0", #str)
assert(util.strcmp(ds_start, zeros) == 0)

-- copy in another dataspace
local ds_copy_in = alloc(ma, #str)

local ds_copy_in_start = attach(rm, ds_copy_in, L4Re.RM_F_RW)
util.strcpy(ds_copy_in_start, str)

ds:access():copy_in(0, ds_copy_in, 0, ds_size)
assert(util.strcmp(ds_start, str) == 0)

print('success')

File: server/src/c/example_dataspace.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "cap_alloc_c.h"
#include "capability_c.h"
#include "consts_c.h"
#include "dataspace_c.h"
#include "env_c.h"
#include "mem_alloc_c.h"
#include "rm_c.h"

char str[] =
  "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod "
  "tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At "
  "vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd "
  "gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum "
  "dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor "
  "invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero "
  "eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no "
  "sea takimata sanctus est Lorem ipsum dolor sit amet.";

#define KOBJ(t) \
  typedef struct kobj_ ## t { \
    l4_cap_idx_t cap; \
    struct l4_re_ ## t obj; \
  } kobj_ ## t;

#define MAKE_KOBJ(t, cap) {cap, l4_cap_ ## t ## _access(cap)}

KOBJ(dataspace)
KOBJ(mem_alloc)
KOBJ(rm)

static kobj_dataspace alloc(kobj_mem_alloc *ma,
                            l4_re_dataspace_size_t size)
{
  l4_re_util_cap_alloc_t cap_alloc = l4_re_util_get_cap_alloc();
  l4_cap_idx_t ds_cap = l4_re_util_counting_cap_alloc_counter_alloc(&cap_alloc);

  kobj_dataspace ds = MAKE_KOBJ(dataspace, ds_cap);

  assert(l4_re_mem_alloc_alloc(&ma->obj,
                               size,
                               ds.cap,
                               0,
                               0) == 0);

  assert(l4_re_dataspace_allocate(&ds.obj, 0, size) == 0);

  return ds;
}

static l4_addr_t reserve(kobj_rm *rm,
                         l4_re_dataspace_size_t size)
{
  l4_addr_t start;
  assert(l4_re_rm_reserve_area(&rm->obj,
                               &start,
                               size,
                               L4_RE_RM_F_SEARCH_ADDR,
                               get_l4_pageshift()) == 0);

  return start;
}

static void map(kobj_dataspace *ds,
                l4_re_dataspace_map_addr_t addr,
                unsigned flags)
{
  assert(l4_re_dataspace_map_region(&ds->obj,
                                    0,
                                    flags,
                                    addr,
                                    addr + l4_re_dataspace_size(&ds->obj)) == 0);
}

static void *attach(kobj_rm *rm,
                    kobj_dataspace *ds,
                    l4_uint_32_t flags)
{
  void *start = NULL;

  assert(l4_re_rm_attach_void(&rm->obj,
                              &start,
                              l4_re_dataspace_size(&ds->obj),
                              L4_RE_RM_F_SEARCH_ADDR | flags,
                              ds->cap,
                              0,
                              get_l4_pageshift()) == 0);

  return start;
}

int main(void)
{
  // environment
  struct l4_re_env env = l4_re_env_env();

  l4_cap_idx_t ma_cap = l4_re_env_mem_alloc_1(&env);
  kobj_mem_alloc ma = MAKE_KOBJ(mem_alloc, ma_cap);

  l4_cap_idx_t rm_cap = l4_re_env_rm_1(&env);
  kobj_rm rm = MAKE_KOBJ(rm, rm_cap);

  // create dataspace
  kobj_dataspace ds = alloc(&ma, sizeof(str));

  l4_re_dataspace_size_t ds_size = l4_re_dataspace_size(&ds.obj);
  assert(ds_size == sizeof(str));

  struct l4_re_dataspace_flags ds_flags = l4_re_dataspace_flags(&ds.obj);
  unsigned long ds_rights = l4_re_dataspace_flags_fpage_rights(&ds_flags);
  assert(ds_rights == L4_RE_DATASPACE_F_RWX);

  // map dataspace
  l4_addr_t start = reserve(&rm, ds_size);

  map(&ds, start, L4_RE_DATASPACE_F_RW);

  memcpy((void *)start, str, ds_size);

  assert(strcmp((char *)start, str) == 0);

  // clear dataspace
  l4_re_dataspace_clear(&ds.obj, 0, ds_size);

  char zeros[sizeof(str)] = {0};
  assert(memcmp((void *)start, zeros, ds_size) == 0);

  // copy in another dataspace
  kobj_dataspace ds_copy_in = alloc(&ma, sizeof(str));

  void *addr_copy_in = attach(&rm, &ds_copy_in, L4_RE_RM_F_RW);
  memcpy(addr_copy_in, str, ds_size);

  l4_re_dataspace_copy_in(&ds.obj, 0, ds_copy_in.cap, 0, ds_size);

  assert(strcmp((char *)start, str) == 0);

  puts("success");

  return EXIT_SUCCESS;
}
